File: haskell/ch1/Ex127.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load ExXXX
-- Run with func_name arg1...
module Ex127 where

import System.Random
import Control.Monad
import Data.Char


fast_times = 32

sq n = n * n

divides a b = (mod b a) == 0

get_rng seed = mkStdGen seed

rand_int :: Int -> StdGen -> (Int, StdGen)
rand_int limit gen =
  let (rawTargetNum, nextGen) = next gen
  in (rem rawTargetNum limit, nextGen)
 
find_divisor n test | sq test > n = n
find_divisor n test | divides test n = test
find_divisor n test = find_divisor n (test + 1)

expmod :: Int -> Int -> Int -> Int
expmod _ 0 _ = 1
expmod base exp m | even exp = rem (sq (expmod base (div exp 2) m)) m
expmod base exp m = rem (base * (expmod base (exp - 1) m)) m

fermat_test :: Int -> Int -> Bool
fermat_test n rnum = 
  let try_it a = (expmod a n n) == a
  in try_it (1 + rnum)

fast_prime n 0 _ = True
fast_prime n times gen =
  let (rnum, newGen) = rand_int (n-1) gen
  in
  if fermat_test n rnum 
  then fast_prime n (times - 1) newGen
  else False

-- Don't know why this is broken
carmichael_num :: Int -> Bool
carmichael_num n =
  let try_it a = (expmod a n n) == a
      iter x = if x == 0 then True
               else if try_it x then iter (n-1)
               else False 
  in and [not $ is_prime n, iter (n-1)]
  
--is_prime n = fast_prime n fast_times (get_rng n)
is_prime n = (find_divisor n 2) == n

test_num n =
  (n,
   is_prime n,
   fast_prime n 128 (get_rng n),
   carmichael_num n)

test_nums vals = report $ map show (map test_num vals)

report vals = mapM_ putStrLn vals

File: javascript/ch1/Ex127.js
// Load with: const ExXXX = require('./ExXXX')
// Run with: ExXXX.xxx()
// Command line: node ExXXX.js

let fast_times = 128;

let sq = n => n * n;

let divides = (a, b) => (b % a) == 0;

let find_divisor = (n, test_divisor) => {
    if (sq(test_divisor) > n) {
        return n;
    } else if (divides(test_divisor, n)) {
        return test_divisor;
    } else {
        return find_divisor(n, test_divisor + 1);
    }
};

let smallest_divisor = n => find_divisor(n, 2);

let even = n => divides(2, n);

let is_prime = n => n == smallest_divisor(n);

let fast_prime = (n, times) => {
    if (times==0) {
        return true;
    } else if (fermat_test(n)) {
        return fast_prime(n, times-1);
    } else {
        return false;
    }
};

let fermat_test = n => {
    let try_it = a => a == expmod(a, n, n);
    return try_it(1 + random(n-1));
}

let carmichael_num = (n) => {
    let car_test = a => a == expmod(a, n, n);
    let iter = x => {
        if (x==0)
            return true;
        else if (car_test(x))
            return iter(x-1);
        else
            return false;
    };
    return !is_prime(n) && iter(n-1);
};

let expmod = (base, exp, m) => {
    if (exp==0) {
        return 1;
    } else if (even(exp)) {
        return sq(expmod(base, exp/2, m)) % m;
    } else {
        return (base * expmod(base, exp-1, m)) % m;
    }
};

let random = limit => Math.floor(limit * Math.random());

let carmichael_result = (n) => {
    console.log("n = " + n + 
                " \tPrime: " + is_prime(n) + 
                " \tFast Prime: " + fast_prime(n, fast_times) + 
                " \tCarmichael: " + carmichael_num(n));
}

exports.carmichael_num = carmichael_num;

exports.test_it = function () {
    carmichael_result(10);
    carmichael_result(11);
    carmichael_result(561);
    carmichael_result(1105);
    carmichael_result(1729);
    carmichael_result(2465);
    carmichael_result(2821);
    carmichael_result(6601);
    //not a carmichael
    carmichael_result(6603);
};

exports.test_it();
